sinks/xray/xray.go

```diff
@@ -0,0 +1,220 @@
+	ParentID    string            `json:"parent_id,omitempty"`
+	StartTime   float64           `json:"start_time"`
+	EndTime     float64           `json:"end_time"`
+	SegmentType string            `json:"type,omitempty"`
+	Namespace   string            `json:"namespace"`
+	Error       bool              `json:"error"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+	Metadata    map[string]string `json:"metadata,omitempty"`
+	HTTP        XRaySegmentHTTP   `json:"http,omitempty"`
+}
+
+// XRaySpanSink is a sink for spans to be sent to AWS X-Ray.
+type XRaySpanSink struct {
+	daemonAddr      string
+	traceClient     *trace.Client
+	conn            *net.UDPConn
+	sampleThreshold uint32
+	commonTags      map[string]string
+	log             *logrus.Logger
+	spansDropped    int64
+	spansHandled    int64
+	nameRegex       *regexp.Regexp
+}
+
+var _ sinks.SpanSink = &XRaySpanSink{}
+
+// NewXRaySpanSink creates a new instance of a XRaySpanSink.
+func NewXRaySpanSink(daemonAddr string, sampleRatePercentage int, commonTags map[string]string, log *logrus.Logger) (*XRaySpanSink, error) {
+
+	log.WithFields(logrus.Fields{
+		"Address": daemonAddr,
+	}).Info("Creating X-Ray client")
+
+	var sampleThreshold uint32
+	if sampleRatePercentage <= 0 || sampleRatePercentage > 100 {
+		return nil, errors.New("Span sample rate percentage must be greater than 0% and less than or equal to 100%")
+	}
+
+	// Set the sample threshold to (sample rate) * (maximum value of uint32), so that
+	// we can store it as a uint32 instead of a float64 and compare apples-to-apples
+	// with the output of our hashing algorithm.
+	sampleThreshold = uint32(sampleRatePercentage * math.MaxUint32 / 100)
+
+	// Build a regex for cleaning names based on valid characters from:
+	// https://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-fields
+	reg, err := regexp.Compile("[^a-zA-Z0-9_\\.\\:\\/\\%\\&#=+\\-\\@\\s\\\\]+")
+	if err != nil {
+		return nil, err
+	}
+
+	return &XRaySpanSink{
+		daemonAddr:      daemonAddr,
+		sampleThreshold: sampleThreshold,
+		commonTags:      commonTags,
+		log:             log,
+		nameRegex:       reg,
+	}, nil
+}
+
+// Start the sink
+func (x *XRaySpanSink) Start(cl *trace.Client) error {
+	x.traceClient = cl
+
+	xrayDaemon, err := net.ResolveUDPAddr("udp", x.daemonAddr)
+	if err != nil {
+		return err
+	}
+	conn, err := net.DialUDP("udp", nil, xrayDaemon)
+	if err != nil {
+		return err
+	}
+	x.conn = conn
+
+	return nil
+}
+
+// Name returns this sink's name.
+func (x *XRaySpanSink) Name() string {
+	return "xray"
+}
+
+// Ingest takes in a span and passed it along to the X-Ray client after
+// some sanity checks and improvements are made.
+func (x *XRaySpanSink) Ingest(ssfSpan *ssf.SSFSpan) error {
+	if err := protocol.ValidateTrace(ssfSpan); err != nil {
+		return err
+	}
+
+	sampleCheckValue := []byte(strconv.FormatInt(ssfSpan.TraceId, 10))
+	hashKey := crc32.ChecksumIEEE(sampleCheckValue)
```

Comment:
Could we add a benchmark for this portion? (Is there a reason we're performing a checksum, instead of doing sampling the way we do for other sinks, like Splunk?)

Reply:
This is pretty old, we should likely just move the sampling over from Splunk to here. Will do!

Reply:
Well, I'm not actually sure! This has some advantages (flexibility and precision), so a benchmark would be good for us to decide which we want to standardize around.

Reply:
So I copied this from the [kafka sink](https://github.com/stripe/veneur/blob/master/sinks/kafka/kafka.go#L335). The implementation is different from the Splunk version. I actually had a hard time reasoning about how the sampling in the Splunk implementation even works. I see now that only settings 1-10 make any sense. The Kafka sink uses a %. We should make these consistent which means maybe this shouldn't change since Splunk is the weird one? :)

I will benchmark them.